### PR TITLE
Set User object as CurrentUser on every request.

### PR DESCRIPTION
**Pull request by kunklejr**
- [x] joeferner
- [x] mwizeman joeybrk372 jharwig sfeng88

We previously made heavy use of the `ProxyUser` class, which prevented heavy-weight objects from being stored in the session and also prevented the loading of a "real" `User` object from the database if all that was required was the userid. However, after some investigation, it turns out
that we almost always load a real `User` object anyway. Additionally, we no longer store anything in the `HttpSession`, so there's really no need for the ProxyUser.

This commit removes the `ProxyUser` and, despite being a large commit, simplifies a lot of things in the process.

Points of Regression:
- login
- authorization

### CODE REVIEW

File: core/core-test/src/main/java/org/visallo/core/user/TestUser.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+    public void setPasswordResetTokenExpirationDate(Date passwordResetTokenExpirationDate) {
+        this.passwordResetTokenExpirationDate = passwordResetTokenExpirationDate;
+    }
+
+    public void setProperties(Map<String, Object> properties) {
+        this.properties = properties;
+    }
+
+    private String username;
```
Comment (mwizeman) [inline core/core-test/src/main/java/org/visallo/core/user/TestUser.java]: Can you re-order the members of this class? Visallo code usually has instance fields first in the class followed by constructors.

File: core/core-test/src/main/java/org/visallo/core/user/TestUser.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+import com.google.common.collect.ImmutableMap;
+import org.json.JSONObject;
+import org.visallo.web.clientapi.model.UserStatus;
+import org.visallo.web.clientapi.model.UserType;
+
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+public class TestUser implements User {
```
Comment (mwizeman) [inline core/core-test/src/main/java/org/visallo/core/user/TestUser.java]: You might be able to save a few lines of code here by making this class extend InMemoryUser.

File: web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java [hunk 3]
```diff
@@ -29,111 +27,88 @@
...
+    @Before
+    public void beforeTest() {
+        when(request.getAttribute(CurrentUser.CURRENT_USER_REQ_ATTR_NAME)).thenReturn(user);
     }
 
     @Test
     public void testGetActiveWorkspaceIdOrDefaultFromAttribute() {
-        ProxyUser user = new ProxyUser(USER_ID, userRepository);
-        when(request.getAttribute(VisalloBaseParameterProvider.USER_REQUEST_ATTRIBUTE_NAME)).thenReturn(user);
+        User user = new TestUser(USER_ID);
```
Comment (mwizeman) [inline web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java]: I think you can remove this line.

File: web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java [hunk 4]
```diff
@@ -29,111 +27,88 @@
...
-        String workspaceId = getActiveWorkspaceIdOrDefault(request, workspaceRepository, userRepository);
+        String workspaceId = getActiveWorkspaceIdOrDefault(request, workspaceRepository);
         assertEquals(WORKSPACE_ID, workspaceId);
     }
 
     @Test
     public void testGetActiveWorkspaceIdOrDefaultFromHeader() {
-        ProxyUser user = new ProxyUser(USER_ID, userRepository);
-        when(request.getAttribute(VisalloBaseParameterProvider.USER_REQUEST_ATTRIBUTE_NAME)).thenReturn(user);
+        User user = new TestUser(USER_ID);
```
Comment (mwizeman) [inline web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java]: I think you can remove this line.

File: web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java [hunk 5]
```diff
@@ -29,111 +27,88 @@
...
-        String workspaceId = getActiveWorkspaceIdOrDefault(request, workspaceRepository, userRepository);
+        String workspaceId = getActiveWorkspaceIdOrDefault(request, workspaceRepository);
         assertEquals(WORKSPACE_ID, workspaceId);
     }
 
     @Test
     public void testGetActiveWorkspaceIdOrDefaultFromParameter() {
-        ProxyUser user = new ProxyUser(USER_ID, userRepository);
-        when(request.getAttribute(VisalloBaseParameterProvider.USER_REQUEST_ATTRIBUTE_NAME)).thenReturn(user);
+        User user = new TestUser(USER_ID);
```
Comment (mwizeman) [inline web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java]: I think you can remove this line.

File: web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java [hunk 6]
```diff
@@ -29,111 +27,88 @@
...
         when(request.getParameter(WORKSPACE_ID_ATTRIBUTE_NAME)).thenReturn(WORKSPACE_ID);
-        getActiveWorkspaceIdOrDefault(request, workspaceRepository, userRepository);
+        getActiveWorkspaceIdOrDefault(request, workspaceRepository);
     }
 
     @Test(expected = VisalloAccessDeniedException.class)
     public void testGetActiveWorkspaceIdOrDefaultWithNoAccess() {
-        ProxyUser user = new ProxyUser(USER_ID, userRepository);
-        when(request.getAttribute(VisalloBaseParameterProvider.USER_REQUEST_ATTRIBUTE_NAME)).thenReturn(user);
+        User user = new TestUser(USER_ID);
```
Comment (mwizeman) [inline web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java]: I think you can remove this line.

File: web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java [hunk 7]
```diff
@@ -29,111 +27,88 @@
...
 
-        getActiveWorkspaceIdOrDefault(request, workspaceRepository, userRepository);
+        getActiveWorkspaceIdOrDefault(request, workspaceRepository);
     }
 
     @Test
     public void testGetActiveWorkspaceId() {
-        ProxyUser user = new ProxyUser(USER_ID, userRepository);
-        when(request.getAttribute(VisalloBaseParameterProvider.USER_REQUEST_ATTRIBUTE_NAME)).thenReturn(user);
+        User user = new TestUser(USER_ID);
```
Comment (mwizeman) [inline web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java]: I think you can remove this line.

File: web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java [hunk 8]
```diff
@@ -29,111 +27,88 @@
...
     public void testGetActiveWorkspaceIdWithMissingWorkspaceId() {
-        getActiveWorkspaceId(request, workspaceRepository, userRepository);
+        getActiveWorkspaceId(request, workspaceRepository);
     }
 
     @Test(expected = VisalloAccessDeniedException.class)
     public void testGetActiveWorkspaceIdNoAccess() {
-        ProxyUser user = new ProxyUser(USER_ID, userRepository);
-        when(request.getAttribute(VisalloBaseParameterProvider.USER_REQUEST_ATTRIBUTE_NAME)).thenReturn(user);
+        User user = new TestUser(USER_ID);
```
Comment (mwizeman) [inline web/web-base-test/src/test/java/org/visallo/web/parameterProviders/VisalloBaseParameterProviderTest.java]: I think you can remove this line.

**Comment by kunklejr** [comment]
@mwizeman I just decided to delete `TestUser` and use the existing `InMemoryUser` with an extra constructor.

File: web/web-base-test/src/test/java/org/visallo/web/parameterProviders/AuthorizationsParameterProviderFactoryTest.java [hunk 9]
```diff
@@ -25,33 +25,25 @@
...
     @Mock
     private WorkspaceRepository workspaceRepository;
 
-    private ProxyUser proxyUser;
-
-    @Before
-    public void before() {
-        proxyUser = new ProxyUser("user123", userRepository);
-    }
+    private User testUser = new InMemoryUser("user123", null, null, null, null);
```
Comment (joeferner) [inline web/web-base-test/src/test/java/org/visallo/web/parameterProviders/AuthorizationsParameterProviderFactoryTest.java]: Looks like this pattern is used a couple times. Should `InMemoryUser` also have a constructor that just takes the user id.

Comment (kunklejr) [reply web/web-base-test/src/test/java/org/visallo/web/parameterProviders/AuthorizationsParameterProviderFactoryTest.java]: I agree. Changing...

**Pull request merged**